### PR TITLE
Implement ChainRPC API

### LINKED ISSUE #171: Implement Chain RPC API

### PR DESCRIPTION
**Pull request by ablax**
- [x] Implement all RPC methods
- [x] Implement all pubsub methods
- [x] Add logic to publish messages concerning chain subscriptions

Closes: #171

### CODE REVIEW

File: src/main/java/com/limechain/rpc/methods/chain/ChainRPCImpl.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+     * - U32: An integer or long value directly representing the block number.
+     * - ARRAY: An array of values, where each value can be either HEX or U32 type.
+     * This method is designed to handle the flexible input types and convert them into a standardized BigInteger format.
+     *
+     * @return A function that takes an Object (HEX, U32, or ARRAY) and returns a BigInteger representation of the block number.
+     * If the input is an array, the function processes each element according to its type (HEX or U32) and returns a list of BigInteger.
+     * Returns null if the input cannot be converted to a BigInteger.
+     */
+    @NotNull
+    private static Function<Object, BigInteger> parseObjectToBigInt() {
```
Comment (georg-getz) [inline src/main/java/com/limechain/rpc/methods/chain/ChainRPCImpl.java]: It's better for private functions to be placed below the public functions they are used in, the top function of a class should never be private.

File: src/main/java/com/limechain/rpc/pubsub/subscriberchannel/Subscriber.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import org.springframework.web.socket.WebSocketSession;
+
+@Data
+@AllArgsConstructor
+public class Subscriber {
+    private final String subscriptionId;
+    private final WebSocketSession session;
+}
```
Comment (georg-getz) [inline src/main/java/com/limechain/rpc/pubsub/subscriberchannel/Subscriber.java]: Data contains `ReqArgCons` which is effectively `AllArgCons` when all the fields are final.

File: src/main/java/com/limechain/rpc/subscriptions/chainsub/ChainSub.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    private static final ChainSub INSTANCE = new ChainSub();
+    private final PubSubService pubSubService = PubSubService.getInstance();
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final BlockState blockState = BlockState.getInstance();
+
+    /**
+     * Private constructor to avoid client applications using the constructor
+     */
+    private ChainSub() {
+    }
```
Comment (georg-getz) [inline src/main/java/com/limechain/rpc/subscriptions/chainsub/ChainSub.java]: Could be a lombok annotation

File: src/main/java/com/limechain/rpc/subscriptions/chainsub/ChainSub.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    public void notifyFinalizedChainHead(final BlockHeader blockHeader) {
+        //CHAIN_FINALIZED_HEAD listens for the new block that is finalized
+        final Message message = createMessageFromHeader(Topic.CHAIN_FINALIZED_HEAD, blockHeader);
+        pubSubService.addMessageToQueue(message);
+
+        pubSubService.broadcast();
+        pubSubService.notifySubscribers();
+    }
+
+
```
Comment (georg-getz) [inline src/main/java/com/limechain/rpc/subscriptions/chainsub/ChainSub.java]: delete one of the lines

File: src/main/java/com/limechain/rpc/config/SubscriptionName.java [hunk 5]
```diff
@@ -14,7 +14,13 @@ public enum SubscriptionName {
...
     CHAIN_HEAD_UNSTABLE_STOP_CALL("chainHead_unstable_stopCall"),
     TRANSACTION_UNSTABLE_SUBMIT_AND_WATCH("transaction_unstable_submitAndWatch"),
-    TRANSACTION_UNSTABLE_UNWATCH("transaction_unstable_unwatch");
+    TRANSACTION_UNSTABLE_UNWATCH("transaction_unstable_unwatch"),
+    CHAIN_SUBSCRIBE_ALL_HEADS("chain_subscribeAllHeads"),
+    CHAIN_UNSUBSCRIBE_ALL_HEADS("chain_unsubscribeAllHeads"),
+    CHAIN_SUBSCRIBE_NEW_HEADS("chain_subscribeNewHeads"),
+    CHAIN_UNSUBSCRIBE_NEW_HEADS("chain_unsubscribeNewHeads"),
+    CHAIN_SUBSCRIBE_FINALIZED_HEADS("chain_subscribeFinalizedHeads"),
+    CHAIN_UNSUBSCRIBE_FINALIZED_HEADS("chain_unsubscribeFinalizedHeads");
```
Comment (vikinatora) [inline src/main/java/com/limechain/rpc/config/SubscriptionName.java:23]: Are these two not indented properly or is it my browser messing thing up?

**Review by georg-getz — APPROVED** [review]

**Pull request merged**